host/xtest/stats.c

```diff
@@ -209,6 +218,98 @@ static int stat_memleak(int argc, char *argv[])
 	return close_sess(&ctx, &sess);
 }
 
+static int stat_loaded_ta(int argc, char *argv[])
+{
+	TEEC_Context ctx = { };
+	TEEC_Session sess = { };
+	TEEC_Result res = TEEC_ERROR_GENERIC;
+	uint32_t eo = 0;
+	TEEC_Operation op = { };
+	struct ta_dump_stats *stats = NULL;
+	size_t stats_size_bytes = 0;
+	size_t n = 0;
+
+	UNUSED(argv);
+	if (argc != 1)
+		return usage();
+
+	open_sess(&ctx, &sess);
+
+	op.paramTypes = TEEC_PARAM_TYPES(TEEC_MEMREF_TEMP_OUTPUT,
+					 TEEC_NONE, TEEC_NONE, TEEC_NONE);
+	res = TEEC_InvokeCommand(&sess, STATS_CMD_TA_STATS, &op, &eo);
+	if (res != TEEC_ERROR_SHORT_BUFFER) {
+		if (res == TEEC_ERROR_NOT_SUPPORTED)
+			printf("Build OP-TEE OS without CFG_TA_STATS ?= y\n");
+		errx(EXIT_FAILURE,
+		     "TEEC_InvokeCommand: res %#"PRIx32" err_orig %#"PRIx32,
+		     res, eo);
+	}
+
+	stats_size_bytes = op.params[0].tmpref.size;
+	if (stats_size_bytes == 0) {
+		printf("not available loaded TA!\n");
+		return close_sess(&ctx, &sess);
+	}
+
+	if (stats_size_bytes % sizeof(*stats))
+		errx(EXIT_FAILURE,
+		     "STATS_CMD_TA_STATS: %zu not a multiple of %zu",
+		     stats_size_bytes, sizeof(*stats));
+	stats = calloc(1, stats_size_bytes);
+	if (!stats)
+		err(EXIT_FAILURE, "calloc(1, %zu)", stats_size_bytes);
+
+	op.params[0].tmpref.buffer = stats;
+	op.params[0].tmpref.size = stats_size_bytes;
+	res = TEEC_InvokeCommand(&sess, STATS_CMD_TA_STATS, &op, &eo);
+	if (res)
+		errx(EXIT_FAILURE,
+		     "TEEC_InvokeCommand: res %#"PRIx32" err_orig %#"PRIx32,
+		     res, eo);
+
+	if (op.params[0].tmpref.size != stats_size_bytes)
```

Comment:
What if a TA was loaded or unloaded in between?  Perhaps it's better to allocate a bit extra and then accept that we might get a bit less. If it still wasn't enough allocate more and try again. It makes the code a bit more complicated but it will improve the user experience in the odd case.

Reply:
Got your point. We can always allocate two more `stats` and if it's still not enough, try 10 times before return error.